### PR TITLE
test: Add basic invariant testing (SC-459)

### PR DESCRIPTION
**Pull request by lucas-manuel**

### CODE REVIEW

File: test/invariant/Invariants.t.sol [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.13;
+
+import "forge-std/Test.sol";
+
+import { PSM3 } from "src/PSM3.sol";
```
Comment (hexonaut) [inline test/invariant/Invariants.t.sol:6]: Unused.

File: test/invariant/handlers/LpHandler.sol [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import { PSM3 } from "src/PSM3.sol";
+
+contract LpHandler is HandlerBase {
+
+    address[] public lps;
+
+    uint256 public depositCount;
+    uint256 public withdrawCount;
+
+    uint256 public constant TRILLION = 1e10;
```
Comment (hexonaut) [inline test/invariant/handlers/LpHandler.sol:17]: 1 trillion is `1e12`.

File: test/invariant/handlers/LpHandler.sol [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    constructor(
+        PSM3      psm_,
+        MockERC20 asset0,
+        MockERC20 asset1,
+        MockERC20 asset2,
+        uint256 lpCount
```
Comment (hexonaut) [inline test/invariant/handlers/LpHandler.sol:24]: Align

File: test/invariant/handlers/LpHandler.sol [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    ) HandlerBase(psm_, asset0, asset1, asset2) {
+        for (uint256 i = 0; i < lpCount; i++) {
+            lps.push(makeAddr(string(abi.encodePacked("lp-", i))));
+        }
+    }
+
+    function _getLP(uint256 lpSeed) internal view returns (address) {
+        return lps[_bound(lpSeed, 0, lps.length - 1)];
```
Comment (hexonaut) [inline test/invariant/handlers/LpHandler.sol:32]: It's actually simpler to use mod division here: `return lps[lpSeed % lps.length];`

File: test/invariant/handlers/SwapperHandler.sol [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    uint256 public swapCount;
+    uint256 public zeroBalanceCount;
+
+    constructor(
+        PSM3      psm_,
+        MockERC20 asset0,
+        MockERC20 asset1,
+        MockERC20 asset2,
+        uint256 lpCount
```
Comment (hexonaut) [inline test/invariant/handlers/SwapperHandler.sol:22]: Align

File: test/invariant/handlers/SwapperHandler.sol [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    ) HandlerBase(psm_, asset0, asset1, asset2) {
+        for (uint256 i = 0; i < lpCount; i++) {
+            swappers.push(makeAddr(string(abi.encodePacked("swapper-", i))));
+        }
+    }
+
+    function _getSwapper(uint256 indexSeed) internal view returns (address) {
+        return swappers[_bound(indexSeed, 0, swappers.length - 1)];
```
Comment (hexonaut) [inline test/invariant/handlers/SwapperHandler.sol:30]: Use mod division.

File: test/invariant/handlers/HandlerBase.sol [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    ) {
+        psm = psm_;
+
+        assets[0] = asset0;
+        assets[1] = asset1;
+        assets[2] = asset2;
+    }
+
+    function _getAsset(uint256 indexSeed) internal view returns (MockERC20) {
+        return assets[_bound(indexSeed, 0, 2)];
```
Comment (hexonaut) [inline test/invariant/handlers/HandlerBase.sol:34]: Mod division.

File: test/invariant/handlers/SwapperHandler.sol [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        if (assetIn == assetOut) {
+            assetOut = _getAsset(assetOutSeed + 2);
+        }
+
+        // By calculating the amount of assetIn we can get from the max asset out, we can
+        // determine the max amount of assetIn we can swap since its the same both ways.
+        uint256 maxAmountIn = psm.previewSwap(
+            address(assetOut),
+            address(assetIn),
+            assetOut.balanceOf(address(psm)
```
Comment (hexonaut) [inline test/invariant/handlers/SwapperHandler.sol:59]: Closing `)` should be on this line instead of next.

**Review by hexonaut — CHANGES_REQUESTED** [review]
Looks good overall. Some minor changes.

File: test/invariant/Invariants.t.sol (hunk 1 above)
Comment (lucas-manuel) [reply test/invariant/Invariants.t.sol]: Removed

File: test/invariant/handlers/LpHandler.sol (hunk 2 above)
Comment (lucas-manuel) [reply test/invariant/handlers/LpHandler.sol]: Ah yep used to be 10 billion fixed

File: test/invariant/handlers/LpHandler.sol (hunk 4 above)
Comment (lucas-manuel) [reply test/invariant/handlers/LpHandler.sol]: Fixed

File: test/invariant/handlers/SwapperHandler.sol (hunk 5 above)
Comment (lucas-manuel) [reply test/invariant/handlers/SwapperHandler.sol]: Fixed

File: test/invariant/handlers/SwapperHandler.sol (hunk 6 above)
Comment (lucas-manuel) [reply test/invariant/handlers/SwapperHandler.sol]: Updated

File: test/invariant/handlers/HandlerBase.sol (hunk 7 above)
Comment (lucas-manuel) [reply test/invariant/handlers/HandlerBase.sol]: Updated

File: test/invariant/handlers/SwapperHandler.sol (hunk 8 above)
Comment (lucas-manuel) [reply test/invariant/handlers/SwapperHandler.sol]: Fixed

**Review by hexonaut — DISMISSED** [review]
LGTM

**Review by hexonaut — APPROVED** [review]

**Pull request merged**